design/ceph-osd-redesign-for-octopus.md

```diff
@@ -0,0 +1,344 @@
+it can be prevented.
+
+This design proposes to change the mental model of where state is "declared" for disk resources.
+Ceph OSDs already keep metadata regarding which Ceph cluster they belong to, and Rook can view this
+metadata as a declaration of state.
+
+Rook users will still wish to add individual disks to their Ceph clusters to be used as OSDs, but
+the deletion-by-omission functionality is removed. Disks which users select for their Ceph clusters
+can be viewed as a declaration of intent to **add** disks as OSDs, but once disks are added, the
+disk content itself becomes the declaration of **existence** of OSDs. Deleting OSDs currently part
+of the Ceph cluster will be a process which requires some manual user intervention via Ceph's CLI or
+dashboard tools.
+
+#### Scheduling OSDs on nodes
+User error is the largest cause of cluster failures, and Rook should also take actions to favor data
+reliability over changes in Kubernetes configurations that might result in OSD pods not being
+scheduled on nodes which contain disks that are part of the current Ceph cluster. Specifically, a
+user might change taints/tolerations, affinities/anti-affinities, or the Cluster CRD such that a
+node previously used for Ceph OSDs is no longer eligible for running OSDs.
+
+This design also proposes to change the mental model of state "declaration" in these cases as well.
+In some ways, node selection methods can be thought of as acting similarly to Kubernetes'
+`RequiredDuringSchedulingIgnoredDuringExecution` affinity. Rook should view node-selection criteria
+as requirements which must be met for provisioning storage from **new** nodes to the Ceph cluster,
+but  Rook should still keep storage which was previously provisioned.
+
+Practically, speaking, Kubernetes keeps the state required to know when to orchestrate
+previously-existing OSDs on nodes which are no longer eligible for provisioning new OSDs. Rook
+should continue to orchestrate existing OSD deployments regardless of their node placement. To do
+this, Rook will need to automatically apply new tolerations to OSD existing deployments if the node
+on which those OSDs are running has a taint added which would prevent the OSD pod from starting.
+
+#### Automatic OSD deletion
+Deletion of OSDs can happen automatically for OSDs which are no longer part of the Ceph cluster. In
+the event of an OSD failure, for example, Ceph will eventually detect that the osd has failed and
+mark it `out`. Once the data for that OSD is rebalanced, Ceph will mark that OSD as being safe to
+delete. When running an orchestration, Rook can safely delete OSDs that are marked safe to delete,
+because at this point, reintroducing the OSD to the Ceph cluster would have no benefits compared to
+creating a new OSD. For die-hard Ceph administrators, this should have an enable/disable switch in
+the CRD to give power users the control they desire.
+
+#### Test environments without disks
```

Comment:
Is it unclear here that this is intended for test environments?

Reply:
I have asked this, and @travisn seems pretty adamant that it's required.

Reply:
If you need more weight on this, then here is mine. I think we should drop osd on file entirely.
I've thought about that too (thanks @BlaineEXE for reminding me) when programming, I realized that I wasn't testing the entire OSD code path to bootstrap OSD because I was using directory. Eventually, I've stopped doing this and added virtual disks to my VM.
Even if you are a developer, you should have the proper setup, same goes for CI.

We already have developers on MacOS and Linux, so we can share our setup and document it so that newcomers will know how to build a proper dev env.